testsuite/MDAnalysisTests/analysis/test_density.py

```diff
@@ -105,6 +112,67 @@ def test_check_set_unit_nolength(self, D):
         with pytest.raises(ValueError):
             D._check_set_unit(units)
 
+# 82 -> 84%; lines 757-758
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        assert D._check_set_unit(units) == None
+
+    # line 774
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        D._check_set_unit(units)
+        del D.units['density']
+        D._check_set_unit({})
+        assert D.units['density'] == None
+
+    # 84 -> 86%; lines 789-792
+    def test_parameters_isdensity(self, D):
+        with pytest.warns(UserWarning, match='Running make_density()'):
+            D.make_density()
+
+    # 86 -> 89%; lines 869-871
+    def test_check_convert_density_grid_not_density(self, D1):
+        unit = 'nm^{-3}'
+        with pytest.raises(RuntimeError, match="The grid is not a density"):
+            D1.convert_density(unit)
+
+    # 89 -> 92%; lines 874-880
+    def test_check_convert_density_value_error(self, D):
+        unit = 'A^{-2}'
+        with pytest.raises(ValueError, match="The name of the unit"):
+            D.convert_density(unit)
+
+    # 92 -> 93%; line 872
+    def test_check_convert_density_units_same_density_units(self, D):
+        unit = 'A^{-3}'
+        D.convert_density(unit)
+        assert unit == D.units['density']
+
+    # 93 -> 93%; line 881
+    def test_check_convert_density_units_density(self, D):
+        unit = 'nm^{-3}'
+        D.convert_density(unit)
+        assert D.units['density'] == unit
+
+    # 93 -> 95%; line 822
+    def test_convert_length_same_length_units(self, D):
+        unit = 'A'
+        D.convert_length(unit)
+        assert unit == D.units['length']
+
+    # 95 -> 97%; line 884
+    def test_parameters_is_Density(self, D, D1):
+        D.__repr__()
+        D1.__repr__()
+        assert 'isDensity' in D1.parameters
+
+    # 97 -> 100%; lines 824-827
+    def test_check_convert_length(sef, D):
```

Comment:
Same as above, we probably should be checking that the edges, origin, etc... are all as expected?

Reply:
In the previous submission I had a question regarding the `numpy_array_almost_equal` to check the elements you mentioned are the same since I was getting the following error:
`numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.`
According to the documentation, `numpy_array_almost_equal` should work for numpy arrays. I got confused and was asking for some guide to approach this.

I think I fixed it by comparing element by element using `zip` as you can see for edges in `test_check_convert_density` and `test_check_convert_length` of the most recent commit (lines 168-169, and 175-176) . Can you please confirm that's a valid approach before adding to origin, etc?
Thanks.

Reply:
If you can give us a example of what you were trying to use the `assert_almost_equal` on (note, if I remember correctly it's best to use that over `array_almost_equal` @lilyminium had some explanation about this somewhere), I can probably offer some insight into what's going on.

For the `grid`, `origin`, and `delta` attributes, assert_almost_equal should work as-is.

For the `edges` attribute you'll probably have to iterate (I can't remember if there's a better way to deal with that kind of list of arrays), something like what you're doing would work.

Reply:
An example of the behaviour is below
```
    def test_check_fail(self, D, h_and_edges):
        h, edges = h_and_edges
        assert_almost_equal(edges, D.edges)
    
    def test_check_no_fail(self, D, h_and_edges):
        h, edges = h_and_edges
        for i, j in zip(edges, D.edges):
            assert_almost_equal(i, j)
```
The error I get is
```
E       numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.

../../../opt/anaconda3/envs/mdanalysis-dev/lib/python3.7/site-packages/numpy/core/_asarray.py:102: VisibleDeprecationWarning
```

The assertion is intended for the `edges`, which is a numpy array. [The documentation](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) gives an example of comparison of two numpy arrays. 
Printing `edges` and `D.edges` reads as
```
[array([ 0.        ,  3.33333333,  6.66666667, 10.        ]), array([ 0. ,  2.5,  5. ,  7.5, 10. ]), array([ 0.,  2.,  4.,  6.,  8., 10.])]
[array([ 0.        ,  3.33333333,  6.66666667, 10.        ]), array([ 0. ,  2.5,  5. ,  7.5, 10. ]), array([ 0.,  2.,  4.,  6.,  8., 10.])]

```
Hence my question if the approach by using `zip` to compare individual elements is valid. I get that in this case `edges` is list of arrays. Which I eventually could transform to `np.array(edges)`, but even in that case `assert_almost_equal` fails and prints the same error.

Reply:
The zip approach is good: the elements of the list are simply arrays of different lengths; I can't think of a more sensible way to do the tests than what you did.